### DISCUSSION TITLE
Gamepad integration

### DISCUSSION THREAD

**Discussion by albinekb** [marc2332/freya#791] [General]
Hello! Amazing library. I got it running on my [RGB30](https://powkiddy.com/products/pre-sale-powkiddy-rgb30-rk3566-handheld-game-console-built-in-wifi) handheld console with little hassle now that it's running the mainline kernel.

I am trying to figure out how I can use the accessibility apis to enable navigation using the gamepad.

I can successfully focus next/prev node using gilrs and use_platform() hook:

```
gilrs::EventType::ButtonPressed(button, code) => {
  match button {
    Button::DPadLeft => {
       let _ = platform.send(EventMessage::FocusPrevAccessibilityNode);
     }
     Button::DPadRight => {
        let _ = platform.send(EventMessage::FocusNextAccessibilityNode);
      }
}
```

Now I am trying to figure out how to get the currently focused nodeId so that I can trigger the click handler:

```
let action = ActionRequest {
    action: accesskit::Action::Default,
    target: nodeId,
    data: None,
};
let window_event =
    accesskit_winit::WindowEvent::ActionRequested(action);
let evt: EventMessage =
    EventMessage::Accessibility(window_event);

let _ = platform.send(evt);
```

Is there some way I can access the `focused_id`?
https://github.com/marc2332/freya/blob/1cd1d2cb16a1f94a8a3633e2287f846f463a92b3/crates/core/src/platform_state.rs#L8

Would it make sense to add it to `use_platform_information` or some other way to access the focused_id?  Perhaps adding it as a PluginEvent?

Or am I doing this from the wrong direction, is there a smarter way to integrate gamepad/inject custom keyboard/accessability events?

Thanks!

**Comment by marc2332** [comment]
Hey! Interesting thing you got in your hands!

You can read whether an element is focused or not with the `use_focus` hook.

Example:

```rs
#[allow(non_snake_case)]
fn CoolComp() -> Element {
    let mut focus = use_focus();
    let is_focused = focus_manager.is_focused();

    rsx!(
        rect {
            width: "100",
            height: "60",
            focus_id: focus.attribute(),
            background: "rgb(45, 45, 45)",
            color: "white",
            label {
                "Am I focused? {is_focused}"
            }
        }
    )
}
```

Not sure what you meant by `so that I can trigger the click handler:`, how is a click event handler related to this? Maybe if the example I shared still didn't work, you could share a bit of your code to give me some context 😄 

> Or am I doing this from the wrong direction, is there a smarter way to integrate gamepad/inject custom keyboard/accessability events?
Honestly, it's hard to tell! But I think that if it works, it works! I never attempted anything similar before, so perhaps there are some rough edges that I could make more flexible, like allowing other types of inputs like game pads. Your implementation reminds me of Freya's `NativeContainer` component which does a similar thing to focus the previous and next elements.

Let me know if that works or if you still need help :)

**Reply by albinekb** [reply]
Thanks! That could work, but I would need to setup `use_focus()` for every button i want to be able to click?

The idea is that I want to use the focus manager to navigate some buttons/lists etc using the gamepad. I can move between the accessible elements using the same code as in `NativeContainer`, however, I cannot find any way to trigger the "onclick" function on the focused element. When I run freya examples on macOS, navigation with tab/shift+tab works, but similarly there's no way to trigger the currently focused element? 

Seems like either event mapping/handling or hooking up the apis and exposing it via a hook would be nice!

For now I can manage the state/navigation manually, I'm mostly just playing around 😄 (trying to make a nice GUI app to record DJ sessions via a soundcard connected to the RGB30 💿 🎵

**Reply by marc2332** [reply]
> Thanks! That could work, but I would need to setup use_focus() for every button i want to be able to click?

Yep, its how you tell Freya that a certain element is focusable.

> but similarly there's no way to trigger the currently focused element?

Some of the components of Freya can be actioned by pressing Enter when focused. This is how its done in button for example: https://github.com/marc2332/freya/blob/1cd1d2cb16a1f94a8a3633e2287f846f463a92b3/crates/components/src/button.rs#L154-L160

If there was an API to map a game pad press to -> keybown event with Enter being pressed perhaps it could work. Maybe as a plugin?

**Reply by marc2332** [reply]
> If there was an API to map a game pad press to -> keybown event with Enter being pressed perhaps it could work. Maybe as a plugin?

I'll try to make POC later today

**Reply by albinekb** [reply]
> Some of the components of Freya can be actioned by pressing Enter when focused. This is how its done in button for example:
> 
> https://github.com/marc2332/freya/blob/1cd1d2cb16a1f94a8a3633e2287f846f463a92b3/crates/components/src/button.rs#L154-L160
> 
> If there was an API to map a game pad press to -> keybown event with Enter being pressed perhaps it could work. Maybe as a plugin?

Yes! I can configure [gptokeyb](https://portmaster.games/gptokeyb-documentation.html) to emulate keyboard events for the gamepad buttons, will try it. I was thinking to use gilrs to get access to analog inputs.

It would be more flexible if I could emit my own `KeyboardEvent` somehow 🤔

**Reply by marc2332** [reply]
👀 https://x.com/mkenzo_8/status/1812901159885283552

**Reply by marc2332** [reply]
https://x.com/mkenzo_8/status/1812908915291791576

https://github.com/marc2332/freya/pull/793/files#diff-5a2fc70e8cb78fa11330bc2c576a533be338d392b208dd8ab2a5aa64f72be41e

What do you think @albinekb ?

**Reply by albinekb** [reply]
Wow @marc2332 👏 Amazing work 🤩  ~Do you have a coffee/beer donation link? 🥇~ Github sponsor ftw 🎉

**Reply by marc2332** [reply]
Thanks for the donation @albinekb, I really appreciate it! 🦀 😄
